src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -1738,6 +1796,61 @@ skip:
   }
   //------------------------------------------------------------------------------------------------------------------------
   template<class t_core>
+  void t_cryptonote_protocol_handler<t_core>::select_peer_for_testing(uint64_t block_height, const cryptonote::blobdata &block, const boost::uuids::uuid &except, std::vector<boost::uuids::uuid> &selected)
+  {
+    std::vector<std::pair<boost::uuids::uuid, int64_t>> peers;
+    m_p2p->for_each_connection([&](cryptonote_connection_context& ctx, nodetool::peerid_type peer_id, uint32_t support_flags)->bool{
+      if (ctx.m_connection_id == except)
+        return true;
+      if (ctx.m_remote_blockchain_height < block_height)
+        return true;
+      if (ctx.m_state < cryptonote_connection_context::state_normal)
+        return true;
+      if (ctx.m_waiting_for_block != "")
+        return true;
+      if (ctx.m_remote_address.get_zone() != epee::net_utils::zone::public_)
+        return true;
+      peers.push_back(std::make_pair(ctx.m_connection_id, ctx.m_score));
+      return true;
+    });
+    if (peers.size() <= 2)
+      return;
+    const size_t n_tests = 1 + peers.size() / 9;
+    const bool sort_by_score = !!(crypto::rand<uint8_t>() & 1);
+    selected.reserve(n_tests);
+    std::shuffle(peers.begin(), peers.end(), crypto::random_device{});
+    if (sort_by_score)
+      std::sort(peers.begin(), peers.end(), [](const std::pair<boost::uuids::uuid, int64_t> &e0, const std::pair<boost::uuids::uuid, int64_t> &e1) {
+        return e0.second < e1.second;
+      });
+    for (size_t idx = 0; idx < n_tests; ++idx)
+    {
+      const auto uuid = peers[idx].first;
+      selected.push_back(uuid);
+      m_p2p->for_connection(uuid, [&](cryptonote_connection_context& ctx, nodetool::peerid_type peer_id, uint32_t f)->bool{
+        ctx.m_waiting_for_block = block;
+        ctx.m_waiting_for_block_deadline = time(NULL) + WAIT_FOR_BLOCK_TIME;
+        MINFO(ctx << "We will be waiting to see if we get block " << block_height << " from peer " << uuid);
+        return true;
+      });
+    }
+  }
+  //------------------------------------------------------------------------------------------------------------------------
+  template<class t_core>
+  void t_cryptonote_protocol_handler<t_core>::check_tested_peer(cryptonote_connection_context& context, const cryptonote::blobdata &block)
+  {
+    if (context.m_waiting_for_block == block)
```

Comment:
Is this racy? Or is the entire context locked before the "handler" for the block arrival? I'll check shortly if you don't remember this off-hand but this has to synchronize that timer somehow.

Reply:
Seems to be.